File: android/proximity/src/main/kotlin/com/hopae/eudi/wallet/android/proximity/Ble.kt
package com.hopae.eudi.wallet.android.proximity

import java.nio.ByteBuffer
import java.util.UUID

/** The ISO/IEC 18013-5 §8.3.3.1.1 mdoc BLE characteristic UUIDs for one mode. */
class BleModeUuids(val state: UUID, val client2Server: UUID, val server2Client: UUID)

/**
 * Shared BLE constants. The GATT server/client transports below are role-agnostic — the mode picks the
 * characteristic set and which side advertises vs scans:
 *  - **peripheral server mode**: mdoc(holder)=GATT server (chars 1/2/3), reader=GATT client.
 *  - **central client mode**: mdoc(holder)=GATT client (chars 5/6/7), reader=GATT server.
 */
object Ble {
    /** Client Characteristic Configuration Descriptor (enables notifications). */
    val CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    /** mdoc peripheral server mode: State / Client2Server / Server2Client = 00000001/2/3. */
    val PERIPHERAL_SERVER = BleModeUuids(mdocUuid(1), mdocUuid(2), mdocUuid(3))

    /** mdoc central client mode: State / Client2Server / Server2Client = 00000005/6/7. */
    val CENTRAL_CLIENT = BleModeUuids(mdocUuid(5), mdocUuid(6), mdocUuid(7))

    /** ISO 18013-5 §8.3.3.1.1.4 Ident characteristic (00000008, Read) — only in the reader's service (central client mode). */
    val IDENT: UUID = mdocUuid(8)

    private fun mdocUuid(n: Int): UUID = UUID.fromString("0000000$n-a123-48ce-896b-4c76973373e6")

    /** 16-byte big-endian encoding of [uuid] (as the DeviceEngagement carries it). */
    fun uuidToBytes(uuid: UUID): ByteArray =
        ByteBuffer.allocate(16).putLong(uuid.mostSignificantBits).putLong(uuid.leastSignificantBits).array()

    fun bytesToUuid(bytes: ByteArray): UUID = ByteBuffer.wrap(bytes).let { UUID(it.long, it.long) }
}

File: apple/Sources/AppleProximity/Ble.swift
import CoreBluetooth
import Foundation

/// ISO/IEC 18013-5 §8.3.3.1.1 BLE constants and framing shared by the peripheral (holder) and central
/// (reader) transports — the iOS counterpart of android `Ble.kt`. Only the fixed mdoc **characteristic**
/// UUIDs live here; the GATT **service** UUID is generated per session by the holder and carried in the
/// QR DeviceEngagement.
enum Ble {
    /// mdoc characteristic UUID template: `0000000{n}-a123-48ce-896b-4c76973373e6` (single-digit n, 1…8).
    static func mdocUuid(_ n: Int) -> CBUUID { CBUUID(string: "0000000\(n)-a123-48ce-896b-4c76973373e6") }

    /// One BLE mode's three data characteristics.
    struct ModeUuids {
        let state: CBUUID
        let client2Server: CBUUID
        let server2Client: CBUUID
    }

    /// mdoc **peripheral server mode** (holder = peripheral): state=1, client2Server=2, server2Client=3.
    static let peripheralServer = ModeUuids(state: mdocUuid(1), client2Server: mdocUuid(2), server2Client: mdocUuid(3))
    /// mdoc **central client mode** (holder = central): state=5, client2Server=6, server2Client=7.
    static let centralClient = ModeUuids(state: mdocUuid(5), client2Server: mdocUuid(6), server2Client: mdocUuid(7))
    /// Ident characteristic (§8.3.3.1.1.4) — read-only, present only in the reader's central-client-mode service.
    static let ident = mdocUuid(8)

    // State characteristic bytes.
    static let stateStart: UInt8 = 0x01 // client → server: begin session (also stops the server advertising)
    static let stateEnd: UInt8 = 0x02   // server → client: end session

    // Data-chunk prefix bytes (client2Server / server2Client).
    static let chunkMore: UInt8 = 0x01 // more chunks follow
    static let chunkLast: UInt8 = 0x00 // final chunk of the message

    /// Inter-chunk gap for paced Write-Without-Response, so each chunk lands in its own connection event and
    /// the receiver peripheral doesn't drop a burst (matches the Android client's pacing).
    static let chunkPacing: TimeInterval = 0.04

    /// Splits a whole framed message into chunks, each prefixed with `chunkMore`/`chunkLast`. `payloadSize`
    /// is the usable bytes per chunk (the negotiated MTU budget minus the 1-byte prefix). An empty message
    /// still yields one final (empty) chunk.
    static func chunk(_ message: [UInt8], payloadSize: Int) -> [[UInt8]] {
        let size = max(payloadSize, 1)
        var chunks: [[UInt8]] = []
        var i = 0
        repeat {
            let end = Swift.min(i + size, message.count)
            let more = end < message.count
            chunks.append([more ? chunkMore : chunkLast] + Array(message[i..<end]))
            i = end
        } while i < message.count
        return chunks
    }

    /// 16-byte big-endian encoding of a UUID, as carried in the DeviceEngagement retrieval method.
    static func uuidBytes(_ uuid: UUID) -> [UInt8] {
        withUnsafeBytes(of: uuid.uuid) { Array($0) }
    }
}
